pkg/yurthub/healthchecker/health_checker.go

```diff
@@ -21,114 +21,368 @@ import (
+func (hcm *healthCheckerManager) backoffEnsureLease(checker *checker) (*coordinationv1.Lease, bool) {
+	var (
+		lease   *coordinationv1.Lease
+		created bool
+		err     error
+	)
+	sleep := 100 * time.Millisecond
+	for {
+		lease, created, err = hcm.ensureLease(checker)
+		if err == nil {
+			break
+		}
+		klog.Warningf("failed to ensure node lease exists, will retry in %v", sleep)
+
+		sleep = sleep * 2
+		if sleep > maxBackoff {
+			klog.Errorf("ensure node lease fail: %v, remoteServer: %v", err, checker.netAddress)
+			return nil, false
+		}
+		hcm.clock.Sleep(sleep)
 	}
+	return lease, created
+}
 
-	if failedRetry == 0 {
-		failedRetry = heartbeatRetry
+func (hcm *healthCheckerManager) ensureLease(checker *checker) (*coordinationv1.Lease, bool, error) {
+	lease, err := checker.leaseClient.Get(hcm.holderIdentity, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		lease, err := checker.leaseClient.Create(hcm.newLease(nil, checker))
+		if err != nil {
+			return nil, false, err
+		}
+		return lease, true, nil
+	} else if err != nil {
+		return nil, false, err
 	}
+	return lease, false, nil
+}
 
-	if healthyThreshold == 0 {
-		healthyThreshold = continuousHealthyCount
+func (hcm *healthCheckerManager) retryUpdateLease(base *coordinationv1.Lease, checker *checker) error {
+	for i := 0; i < hcm.failedRetry; i++ {
+		lease, err := checker.leaseClient.Update(hcm.newLease(base, checker))
+		if err == nil {
+			hcm.setLatestLease(lease)
+			return nil
+		}
+		if apierrors.IsConflict(err) {
+			base, _ = hcm.backoffEnsureLease(checker)
+			continue
+		}
 	}
+	return fmt.Errorf("failed %d attempts to update node lease", hcm.failedRetry)
+}
 
-	c := &checker{
-		serverHealthzAddr: serverHealthzURL.String(),
-		healthzClient:     tp.HealthzHTTPClient(),
-		clusterHealthy:    false,
-		lastTime:          time.Now(),
-		onFailureFunc:     tp.Close,
-		netAddress:        url.Host,
-		failedRetry:       failedRetry,
-		healthyThreshold:  healthyThreshold,
+func (hcm *healthCheckerManager) newLease(base *coordinationv1.Lease, checker *checker) *coordinationv1.Lease {
+	var lease *coordinationv1.Lease
+	if base == nil {
+		lease = &coordinationv1.Lease{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      hcm.holderIdentity,
+				Namespace: corev1.NamespaceNodeLease,
+			},
+			Spec: coordinationv1.LeaseSpec{
+				HolderIdentity:       pointer.StringPtr(hcm.holderIdentity),
+				LeaseDurationSeconds: pointer.Int32Ptr(hcm.leaseDurationSeconds),
+			},
+		}
+	} else {
+		lease = base.DeepCopy()
 	}
 
-	initHealthyStatus, err := PingClusterHealthz(c.healthzClient, c.serverHealthzAddr)
-	if err != nil {
-		klog.Errorf("cluster(%s) init status: unhealthy, %v", c.serverHealthzAddr, err)
+	lease.Spec.RenewTime = &metav1.MicroTime{Time: hcm.clock.Now()}
+
+	if lease.OwnerReferences == nil || len(lease.OwnerReferences) == 0 {
+		if node, err := checker.client.CoreV1().Nodes().Get(hcm.holderIdentity, metav1.GetOptions{}); err == nil {
```

Comment:
can we get node.UID from yurt-hub env instead of getting from kube-apiserver?

Reply:
the env of yurt-hub have not node.UID.